pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/SpoolIntegrationTest.java

```diff
@@ -117,13 +121,146 @@ public void intermediateSpool()
+        + "             SELECT deviceOS,\n"
+        + "                    PERCENTILETDigest(totalTrips, 20) AS p20\n"
+        + "             FROM q7\n"
+        + "             GROUP BY deviceOS\n"
+        + "         ) q ON sd.deviceOS = q.deviceOS\n"
+        + "     )\n"
+        + "SELECT *\n"
+        + "FROM q8");
+    JsonNode stats = jsonNode.get("stageStats");
+    assertNoError(jsonNode);
+    DocumentContext parsed = JsonPath.parse(stats.toString());
+
+    checkSpoolTimes(parsed, 6, 5, 1);
+    checkSpoolTimes(parsed, 6, 14, 1);
+    checkSpoolSame(parsed, 6, 5, 14);
+
+    checkSpoolTimes(parsed, 7, 6, 2);
+
+    checkSpoolTimes(parsed, 4, 3, 1);
+    checkSpoolTimes(parsed, 4, 7, 2); // because there are 2 copies of 7 as well
+    checkSpoolTimes(parsed, 4, 9, 1);
+    checkSpoolTimes(parsed, 4, 12, 1);
+    checkSpoolTimes(parsed, 4, 18, 1);
+    checkSpoolSame(parsed, 4, 3, 7, 9, 12, 18);
+  }
+
+  private List<Map<String, Object>> findDescendantById(DocumentContext stats, int parent, int descendant) {
+    return stats.read(parentDescendantJsonPathExpression(parent, descendant));
+  }
+
+  private void checkSpoolTimes(DocumentContext stats, int spoolStageId, int parent, int times) {
+    List<Map<String, Object>> descendants = findDescendantById(stats, parent, spoolStageId);
+    Assert.assertEquals(descendants.size(), times, "Stage " + spoolStageId + " should be descended from stage "
+        + parent + " exactly " + times + " times");
+    Map<String, Object> firstSpool = descendants.get(0);
+    for (int i = 1; i < descendants.size(); i++) {
+      Assert.assertEquals(descendants.get(i), firstSpool, "Stage " + spoolStageId + " should be the same in "
+          + "all " + times + " descendants");
+    }
+  }
+
+  private void checkSpoolSame(DocumentContext stats, int spoolStageId, int... parents) {
+    List<Pair<Integer, List<Map<String, Object>>>> spools = Arrays.stream(parents)
+        .mapToObj(parent -> Pair.of(parent, findDescendantById(stats, parent, spoolStageId)))
+        .collect(Collectors.toList());
+    Pair<Integer, List<Map<String, Object>>> notEmpty = spools.stream()
+        .filter(s -> !s.getValue().isEmpty())
+        .findFirst()
+        .orElse(null);
+    if (notEmpty == null) {
+      Assert.fail("None of the parent nodes " + Arrays.toString(parents) + " have a descendant with id "
+          + spoolStageId);
+    }
+    List<Pair<Integer, List<Map<String, Object>>>> allNotEqual = spools.stream()
+        .filter(s -> !s.getValue().get(0).equals(notEmpty.getValue().get(0)))
+        .collect(Collectors.toList());
+    if (!allNotEqual.isEmpty()) {
+      Assert.fail("The descendant with id " + spoolStageId + " is not the same in all parent nodes "
+          + spools);
+    }
+  }
 
-    Assert.assertEquals(stage4On3, stage4On7, "Stage 4 should be the same in both stage 3 and stage 7");
+  @Language("jsonpath")
+  private String parentDescendantJsonPathExpression(int parent, int child) {
```

Comment:
I'm not sure I follow what this method is doing and how / why?

Reply:
Oh I missed that this is the same expression, I think the method name threw me off.

> give me all objects whose stage id is child, and one of its descendants (not just direct parent, any ancestor) is parent.

This statement is confusing me - descendants (children + children of children) and ancestors (parents + parents of parents) are the opposite right? So this method is essentially just trying to find all objects where both the specified parent and child stages exist in the expected order?